ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,221 @@
+        } else {
+            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+                log.error("Remote server returned error code: {}", statusCode);
+                tensors = processErrorResponse(statusCode, body);
+            } else {
+                try {
+                    tensors = processOutput(body, connector, scriptService, parameters);
+                } catch (Exception e) {
+                    log.error("Failed to process response body: {}", body, e);
+                    tensors = processErrorResponse(statusCode, body);
+                }
+            }
+        }
+        tensors.setStatusCode(statusCode);
+        tensorOutputs.put(countDownLatch.getSequence(), tensors);
+    }
+
+    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
+        List<ModelTensors> modelTensors = new ArrayList<>();
+        TreeMap<Integer, ModelTensors> sortedMap = new TreeMap<>(tensorOutputs);
+        log.debug("Reordered tensor outputs size is {}", sortedMap.size());
+        if (tensorOutputs.size() == 1) {
+            // batch API case
+            int status = tensorOutputs.get(0).getStatusCode();
+            if (status == HttpStatus.SC_OK) {
+                modelTensors.add(tensorOutputs.get(0));
+                actionListener.onResponse(modelTensors);
+            } else {
+                try {
+                    actionListener
+                        .onFailure(
+                            new OpenSearchStatusException(
+                                AccessController
+                                    .doPrivileged(
+                                        (PrivilegedExceptionAction<String>) () -> GSON
+                                            .toJson(tensorOutputs.get(0).getMlModelTensors().get(0).getDataAsMap())
+                                    ),
+                                RestStatus.fromCode(status)
+                            )
+                        );
+                } catch (PrivilegedActionException e) {
+                    actionListener.onFailure(new OpenSearchStatusException(e.getMessage(), RestStatus.fromCode(statusCode)));
+                }
+            }
+        } else {
+            // non batch API.
+            for (Map.Entry<Integer, ModelTensors> entry : sortedMap.entrySet()) {
+                modelTensors.add(entry.getKey(), entry.getValue());
+            }
+            actionListener.onResponse(modelTensors);
+        }
+    }
+
+    protected class MLResponseSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            s.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(ByteBuffer byteBuffer) {
+            responseBody.append(StandardCharsets.UTF_8.decode(byteBuffer));
```

Comment:
Should we make this default charsets or this can be user dependent?

Reply:
We didn't make this user dependent in current code: https://github.com/opensearch-project/ml-commons/blob/a0358e26f2bc5c9be800fe05ce9938b41142edb5/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/AwsConnectorExecutor.java#L110, and in network transport case, UTF8 is the standard encoding choice I think we can make it default.